windows/src/desktop/kmshell/main/BackgroundUpdate.pas

```diff
@@ -0,0 +1,1192 @@
+begin
+  DownloadUpdate := TDownloadUpdate.Create;
+  try
+    DownloadResult := DownloadUpdate.DownloadUpdates;
+    KL.Log('TBackgroundUpdate.DownloadUpdatesBackground: DownloadResult = '+IntToStr(Ord(DownloadResult)));
+    Result := DownloadResult;
+// #TODO: #10210 workout when we need to refresh kmcom keyboards
+
+//      if Result in [ wucSuccess] then
+//  begin
+//    kmcom.Keyboards.Refresh;
+//    kmcom.Keyboards.Apply;
+//    kmcom.Packages.Refresh;
+//  end;
+
+  finally
+    DownloadUpdate.Free;
+  end;
+end;
+
+{ WaitingRestartState }
+
+procedure WaitingRestartState.Enter;
+begin
+  // Enter DownloadingState
+  bucStateContext.SetRegistryState(usWaitingRestart);
+end;
+
+procedure WaitingRestartState.Exit;
+begin
+  // Exit DownloadingState
+end;
+
+procedure WaitingRestartState.HandleCheck;
+begin
+
+end;
+
+procedure WaitingRestartState.HandleDownload;
+begin
+
+end;
+
+function WaitingRestartState.HandleKmShell;
+var
+  SavedPath : String;
+  Filenames : TStringDynArray;
+begin
+  KL.Log('WaitingRestartState.HandleKmShell Enter');
+  // Still can't go if keyman has run
+  if HasKeymanRun then
+  begin
+    KL.Log('WaitingRestartState.HandleKmShell Keyman Has Run');
+    Result := kmShellExit;
+    // Exit; // Exit is not wokring for some reason.
+    // this else is only here because the exit is not working.
+  end
+  else
+  begin
+    // Check downloaded cache if available then
+    SavedPath := IncludeTrailingPathDelimiter(TKeymanPaths.KeymanUpdateCachePath);
+    GetFileNamesInDirectory(SavedPath, FileNames);
+    if Length(FileNames) = 0 then
+    begin
+        KL.Log('WaitingRestartState.HandleKmShell No Files in Download Cache');
+        // Return to Idle state and check for Updates state
+        ChangeState(IdleState);
+        bucStateContext.CurrentState.HandleCheck;
+        Result := kmShellExit;
+        // Exit; // again exit was not working
+    end
+    else
+    begin
+      KL.Log('WaitingRestartState.HandleKmShell is good to install');
+      ChangeState(InstallingState);
+      Result := kmShellExit;
+    end;
+  end;
+end;
+
+procedure WaitingRestartState.HandleInstall;
+begin
+
+end;
+
+procedure WaitingRestartState.HandleMSIInstallComplete;
+begin
+
+end;
+
+procedure WaitingRestartState.HandleAbort;
+begin
+
+end;
+
+function WaitingRestartState.StateName;
+begin
+
+  Result := 'WaitingRestartState';
+end;
+
+{ InstallingState }
+
+function InstallingState.DoInstallPackage(Package: TBackgroundUpdateParamsPackage): Boolean;
+var
+  FPackage: IKeymanPackageFile2;
+begin
+  Result := True;
+
+  FPackage := kmcom.Packages.GetPackageFromFile(Package.SavePath) as IKeymanPackageFile2;
+  FPackage.Install2(True);  // Force overwrites existing package and leaves most settings for it intact
+  FPackage := nil;
+
+  kmcom.Refresh;
+  kmcom.Apply;
+  System.SysUtils.DeleteFile(Package.SavePath);
+end;
+
+procedure InstallingState.DoInstallKeyman;
+var
+  s: string;
+  FResult: Boolean;
+begin
```

Comment:
Calling `DoInstallKeyman(bucStateContext.FParams.Keyman.SavePath)` would be less WET.